src/compiler/program.ts

```diff
@@ -450,57 +587,83 @@ namespace ts {
                 processSourceFile(normalizePath(referencedFileName), /* isDefaultLib */ false, file, ref.pos, ref.end);
             });
         }
-
-        function processImportedModules(file: SourceFile, basePath: string) {
-            forEach(file.statements, node => {
-                if (node.kind === SyntaxKind.ImportDeclaration || node.kind === SyntaxKind.ImportEqualsDeclaration || node.kind === SyntaxKind.ExportDeclaration) {
-                    let moduleNameExpr = getExternalModuleName(node);
-                    if (moduleNameExpr && moduleNameExpr.kind === SyntaxKind.StringLiteral) {
-                        let moduleNameText = (<LiteralExpression>moduleNameExpr).text;
-                        if (moduleNameText) {
-                            let searchPath = basePath;
-                            let searchName: string; 
-                            while (true) {
-                                searchName = normalizePath(combinePaths(searchPath, moduleNameText));
-                                if (forEach(supportedExtensions, extension => findModuleSourceFile(searchName + extension, moduleNameExpr))) {
-                                    break;
-                                }
-                                let parentPath = getDirectoryPath(searchPath);
-                                if (parentPath === searchPath) {
-                                    break;
+        
+        function processImportedModules(file: SourceFile, basePath: string) {            
+            collectExternalModuleReferences(file);            
+            if (file.imports.length) {               
+                let allImportsInTheCache = false;
+
+                // try to grab existing module resolutions from the old source file
+                let oldSourceFile: SourceFile = oldProgram && oldProgram.getSourceFile(file.fileName);

```

Comment:
why do we need to this here. can not we do this in tryReuseStructureFromOldProgram, just after we got the newSourceFile, we know the old file resolutions, we can do that there, so we have two helpers, processImports, and processImportsIncrementally.. this way the oldProgram, oldSoruceFile knowledge is kept in one place.